File: app/javascript/channels/room_channel.js
import consumer from "./consumer"

$(document).on("turbolinks:load", function() {
  const element = document.querySelector("#words");
  const chatChannel = consumer.subscriptions.create({ channel: "RoomChannel", room: $("#words").data("room_id") }, {
    connected() {
      // Called when the subscription is ready for use on the server
    },

    disconnected() {
      // Called when the subscription has been terminated by the server
    },

    received(data) {
      const received_element = document.createElement("div");
      received_element.className = "word";
      received_element.innerHTML = `
        <p>
        ${data["user_name"]}: ${data["content"]}
        </p>
      `;

      element.appendChild(received_element);
    },

    speak(word) {
      return this.perform('speak', {
        word: word
      });
    }
  });


  $(document).on("keypress", "[data-behavior~=room_speaker]", function(event) {
    if (event.keyCode === 13) {
      chatChannel.speak(event.target.value);
      event.target.value = "";
      return event.preventDefault();
    };
  });
});

File: app/channels/room_channel.rb
class RoomChannel < ApplicationCable::Channel
  def subscribed
    room = Room.find(params["room"])
    stream_for room
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def speak(data)
    if !(have_membership?)
      return
    end
    set_room
    if data["word"] =~ /[^ぁ-んー]/
      send_message "ひらがなのみを使用してください"
    elsif data["word"][-1] == "ん"
      send_message "語尾に「ん」がついてはいけません"
    elsif data["word"] =~ /ー{2,}/
      send_message "「ー」が二つ以上重なってはいけません"
    elsif @room.words.exists?
      @last_word_content = @room.words.last.content
      if @last_word_content[-1] != data["word"][0]
        if @last_word_content[-1] == "ー"
          if @last_word_content[-2] == data["word"][0]
            post_word data["word"], params["room"]
            return
          else
            send_message "前の言葉が「ー」で終わる場合はその前の文字からはじめてください"
          end
        end
        send_message "前の言葉の最後の文字ととこの言葉の最初の文字が一致していません"
      else
        post_word data["word"], params["room"]
      end
    else
      post_word data["word"], params["room"]
    end
  end


  private

  def set_room
    @room = Room.find(params["room"])
  end

  def send_message(message)
    UserMessageBroadcastJob.perform_later message, current_user
  end

  def post_word(word, room_id)
    Word.create! content: word, user_id: current_user.id, room_id: room_id
  end

  def have_membership?
    if Room.find(params["room"]).room_memberships.find_by(user_id: current_user.id)
      true
    else
      false
    end
  end
end
